### PR TITLE
Adds Groovyscript support for crafting

### PR DESCRIPTION
**Pull request by yor42**
## 📝 Description

This PR Adds Groovyscript support for crafting such as ammo press and workbench.

## 🎯 Goals

- add Groovyscript support for crafting such as ammo press and workbench for easier modpack development.

## ❌ Non Goals

- Completely replacing existing json recipe is not a goal for compatibility's sake

## 🚦 Testing 

use following example script on Postinit Script of Groovyscript (E.G: run/groovy/postinit/main.groovy)
```groovy
mods.mwc.craftingstation.removeAllinGroup("GUN")
mods.mwc.craftingstation.recipeBuilder().output(item('minecraft:clay')).addEntry(item('minecraft:nether_star')*4).register()
mods.mwc.craftingstation.recipeBuilder().output(item('minecraft:iron_ingot')).addOredict("oreIron", 4, 0.5).addOredict("oreCoal", 4).register()
```

this script should do 3 things:
1. clear ALL the recipes in gun tab of workbench
2. add 2 example script. 1 recipe that turns 4 nether star into 1 clay block with 100% yield, and 1 recipe that turns 4 Iron Ore (50% Yield) and 1 Coal Ore(100% Yield) (both are ore dictionary entry)  into 1 Iron ingot.

image

image

Test video to prove that this PR doesn't break anything even if Groovyscript isn't present in runtime:

https://github.com/Cubed-Development/Modern-Warfare-Cubed/assets/49832212/6f451206-dbaa-49a1-9c3b-27fc6729c651

## ⏮️ Backwards Compatibility 

Exisiting JSON Recipe system is not replaced. since it should not cause issue with any existing modpack that utilizes the JSON for their costom recipes.

## 📚 Related Issues & Documents

N/A

## 🖼️ Screenshots/Recordings

image

image

## 📖 Added to documentation?

- [ ] 📜 README.md
- [ ] 📑 Documentation
- [x] 📓 Javadoc
- [ ] 🍕 Comments
- [ ] 🙅 No documentation needed

## 😄 [optional] What gif best describes this PR or how it makes you feel?
N/A

## Summary by CodeRabbit

- **New Features**
  - Introduced a new crafting system with enhanced recipe management and crafting stations.
  - Added compatibility with Groovy scripting for mod customization.

- **Enhancements**
  - Updated user interfaces for crafting stations, improving user experience.
  - Streamlined crafting processes, allowing for more intuitive item crafting.

- **Bug Fixes**
  - Resolved issues with item crafting logic that affected the correct item output.
  - Fixed tooltip rendering issues in crafting interfaces.

- **Refactor**
  - Standardized item return types across various methods for consistency.
  - Optimized conditional checks and control flow for better performance.

- **Documentation**
  - Updated in-game tooltips and help sections to reflect new crafting features.

- **Chores**
  - Cleaned up unused import statements and redundant code.

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -69,6 +76,7 @@ repositories {
 dependencies {
     implementation 'Red-Studio-Ragnarok:Red-Core:0.4-Dev-2'
     sources 'Red-Studio-Ragnarok:Red-Core:0.4-Dev-2:sources@jar'
+    compileOnly rfg.deobf("curse.maven:groovyscript-687577:4905039")
```
Comment (Desoroxxx) [inline build.gradle:79]: This should be replaced by properly depending on it
```suggestion
    implementation 'com.cleanroommc:groovyscript:0.7.2'
```

File: build.gradle [hunk 2]
```diff
@@ -55,6 +55,13 @@ repositories {
         name 'Cleanroom'
         url 'https://maven.cleanroommc.com'
     }
+    maven {
+        name 'CurseMaven'
+        url 'https://cursemaven.com'
+        content {
+            includeGroup 'curse.maven'
+        }
+    }
```
Comment (Desoroxxx) [inline build.gradle:64]: This is not needed as now we will use the Cleanroom Maven for GS
```suggestion
```

**Review by Desoroxxx — CHANGES_REQUESTED** [review]
Groovy Script is now properly on the Cleanroom maven, no need to deobfuscate it anymore.

File: src/main/java/com/paneedah/mwc/MWC.java [hunk 3]
```diff
@@ -65,6 +62,7 @@ public final class MWC {
     @SidedProxy(serverSide = "com.paneedah.mwc.proxies.CommonProxy", clientSide = "com.paneedah.mwc.proxies.ClientProxy")
     public static CommonProxy commonProxy;
 
+
     @Mod.EventHandler
```
Comment (Desoroxxx) [inline src/main/java/com/paneedah/mwc/MWC.java:66]: Most likely accidental white space
```suggestion
    public static CommonProxy commonProxy;

    @Mod.EventHandler
```

**Review by Desoroxxx — CHANGES_REQUESTED** [review]

**Review by Desoroxxx — DISMISSED** [review]

**Review by Desoroxxx — APPROVED** [review]

**Comment by Desoroxxx** [comment]
@coderabbitai review

File: src/main/java/com/paneedah/weaponlib/HighIQSpawnEgg.java [hunk 4]
```diff
@@ -173,8 +174,8 @@ public CraftingEntry[] getModernRecipe() {
 	}
 
 	@Override
-	public Item getItem() {
-		return this;
+	public ItemStack getItem() {
```
Comment (Desoroxxx) [inline src/main/java/com/paneedah/weaponlib/HighIQSpawnEgg.java:177]: This should be renamed to `getItemStack`

File: src/main/java/com/paneedah/weaponlib/config/BalancePackManager.java [hunk 5]
```diff
@@ -750,7 +750,7 @@ public static void loadBalancePack(ICommandSender sender, String fileName) {
 			return;
 		}
 
-		JsonObject object = null;
+		JsonObject object;
```
[the remark this answers is not in the document]
Comment (yor42) [reply src/main/java/com/paneedah/weaponlib/config/BalancePackManager.java:753]: nope

**Review by Desoroxxx — APPROVED** [review]

**Pull request merged**